x/merkledb/trieview.go

```diff
@@ -628,81 +626,62 @@ func (t *trieView) commitToDB(ctx context.Context) error {
 	}
 
 	// now commit the parent trie to the db
-	return t.getParentTrie().commitToDB(ctx)
-}
-
-// Assumes [t.validityTrackingLock] isn't held.
-func (t *trieView) isInvalid() bool {
-	t.validityTrackingLock.RLock()
-	defer t.validityTrackingLock.RUnlock()
-
-	return t.invalidated
+	return t.parentTrie.Get().commitToDB(ctx)
 }
 
 // Invalidates this view and all descendants.
-// Assumes [t.validityTrackingLock] isn't held.
+// Assumes [t.childViewsLock] isn't held.
 func (t *trieView) invalidate() {
-	t.validityTrackingLock.Lock()
-	defer t.validityTrackingLock.Unlock()
-
-	t.invalidated = true
-
-	for _, childView := range t.childViews {
-		childView.invalidate()
-	}
+	t.invalidated.Set(true)
 
-	// after invalidating the children, they no longer need to be tracked
-	t.childViews = make([]*trieView, 0, defaultPreallocationSize)
+	t.invalidateChildren()
 }
 
 // Invalidates all children of this view.
-// Assumes [t.validityTrackingLock] isn't held.
+// Assumes [t.childViewsLock] isn't held.
 func (t *trieView) invalidateChildren() {
 	t.invalidateChildrenExcept(nil)
 }
 
 // moveChildViewsToView removes any child views from the trieToCommit and moves them to the current trie view
+// Assumes [t.childViewsLock] isn't held.
 func (t *trieView) moveChildViewsToView(trieToCommit *trieView) {
-	t.validityTrackingLock.Lock()
-	defer t.validityTrackingLock.Unlock()
+	t.childViewsLock.Lock()
+	defer t.childViewsLock.Unlock()
 
-	trieToCommit.validityTrackingLock.Lock()
-	defer trieToCommit.validityTrackingLock.Unlock()
+	trieToCommit.childViewsLock.Lock()
+	defer trieToCommit.childViewsLock.Unlock()
 
 	for _, childView := range trieToCommit.childViews {
-		childView.updateParent(t)
+		childView.parentTrie.Set(t)
 		t.childViews = append(t.childViews, childView)
 	}
-	trieToCommit.childViews = make([]*trieView, 0, defaultPreallocationSize)
-}
-
-func (t *trieView) updateParent(newParent TrieView) {
-	t.validityTrackingLock.Lock()
-	defer t.validityTrackingLock.Unlock()
 
-	t.parentTrie = newParent
+	// clear trieToCommit's children
+	trieToCommit.childViews = make([]*trieView, 0, defaultPreallocationSize)
 }
 
 // Invalidates all children of this view except [exception].
 // [t.childViews] will only contain the exception after invalidation is complete.
-// Assumes [t.validityTrackingLock] isn't held.
+// Assumes [t.childViewsLock] isn't held.
 func (t *trieView) invalidateChildrenExcept(exception *trieView) {
-	t.validityTrackingLock.Lock()
-	defer t.validityTrackingLock.Unlock()
-
-	for _, childView := range t.childViews {
-		if childView != exception {
-			childView.invalidate()
-		}
-	}
+	t.childViewsLock.Lock()
+	invalidChildrenViews := t.childViews
```

Comment:
grabbing the current list and then overwriting it, lets us invalidate the children while we no longer hold the local childViewsLock.